wpilibj/src/main/java/edu/wpi/first/wpilibj/Watchdog.java

```diff
@@ -101,25 +161,100 @@ public void reset() {
     m_epochs.clear();
-    m_notifier.startPeriodic(m_timeout);
+
+    m_queueMutex.lock();
+    try {
+      m_isExpired = false;
+
+      m_watchdogs.remove(this);
+      m_expirationTime = m_startTime + m_timeout;
+      m_watchdogs.add(this);
+      m_schedulerWaiter.signalAll();
+    } finally {
+      m_queueMutex.unlock();
+    }
   }
 
   /**
    * Disable the watchdog.
    */
   public void disable() {
-    m_notifier.stop();
+    m_queueMutex.lock();
+    try {
+      m_isExpired = false;
+
+      m_watchdogs.remove(this);
+      m_schedulerWaiter.signalAll();
+    } finally {
+      m_queueMutex.unlock();
+    }
+  }
+
+  private static Thread startDaemonThread(Runnable target) {
+    Thread inst = new Thread(target);
+    inst.setDaemon(true);
+    inst.start();
+    return inst;
   }
 
-  private void timeoutFunc() {
-    if (!m_isExpired) {
-      System.out.println("Watchdog not fed after " + m_timeout + "s");
-      m_callback.run();
-      m_isExpired = true;
-      disable();
+
+  private static void schedulerFunc() {
+    m_queueMutex.lock();
+
+    try {
+      while (true) {
+        if (m_watchdogs.size() > 0) {
+          boolean timedOut = !awaitUntil(m_schedulerWaiter, m_watchdogs.peek().m_expirationTime);
+          if (timedOut) {
+            if (m_watchdogs.size() == 0 || m_watchdogs.peek().m_expirationTime
+                > RobotController.getFPGATime()) {
+              continue;
+            }
+
+            // If the condition variable timed out, that means a Watchdog timeout
+            // has occurred, so call its timeout function.
+            Watchdog watchdog = m_watchdogs.poll();
+
+            System.out.format("Watchdog not fed within %.6fs\n", watchdog.m_timeout / 1.0e6);
+            m_queueMutex.unlock();
+            watchdog.m_callback.run();
+            m_queueMutex.lock();
+            watchdog.m_isExpired = true;
+          }
+          // Otherwise, a Watchdog removed itself from the queue (it notifies
+          // the scheduler of this) or a spurious wakeup occurred, so just
+          // rewait with the soonest watchdog timeout.
+        } else {
+          while (m_watchdogs.size() == 0) {
+            m_schedulerWaiter.awaitUninterruptibly();
+          }
+        }
+      }
+    } finally {
+      m_queueMutex.unlock();
     }
   }
+
+  /**
+   * Wrapper emulating functionality of C++'s std::condition_variable::wait_until().
+   *
+   * @param cond The condition variable on which to wait.
+   * @param time The time at which to stop waiting.
+   * @return False if the deadline has elapsed upon return, else true.
+   */
+  private static boolean awaitUntil(Condition cond, long time) {
+    long delta = time - RobotController.getFPGATime();
+    try {
+      if (delta > 0) {
+        return cond.await(delta, TimeUnit.MICROSECONDS);
+      }
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      ex.printStackTrace();
+    }
+
+    return true;
```

Comment:
There should be an else branch that returns false. Returning true at the end if the cond var is interrupted should be correct.

Reply:
Would returning false also cause a mutex unlock then lock? We noticed that was the behavior of c++ wait_until if it received a negative timeout.

Reply:
It wouldn't, but it shouldn't need to because the main loop will remove the Watchdog from the queue via PriorityQueue.poll() on timeout. It's a longer wait than otherwise for member functions waiting on the mutex though. Ideally, java's Condition.await() will lock then unlock immediately when given a negative duration. I haven't tested it.

Reply:
https://github.com/wpilibsuite/allwpilib/pull/1499